javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll

```diff
@@ -0,0 +1,285 @@
+/**
+ * Provides default sources, sinks and sanitisers for reasoning about
+ * DOS attacks using objects with unbounded length property. 
+ * As well as extension points for adding your own.
+ */
+
+import javascript
+
+module TaintedLength {
+  import semmle.javascript.security.dataflow.RemoteFlowSources
+  import semmle.javascript.security.TaintedObject
+  import DataFlow::PathGraph
+
+  /**
+   * Holds if an exception will be thrown whenever `e` evaluates to `undefined` or `null`.
+   */
+  predicate isCrashingWithNullValues(Expr e) {
+    exists(ExprOrStmt ctx |
+      e = ctx.(PropAccess).getBase()
+      or
+      e = ctx.(InvokeExpr).getCallee()
+      or
+      e = ctx.(AssignExpr).getRhs() and
+      ctx.(AssignExpr).getLhs() instanceof DestructuringPattern
+      or
+      e = ctx.(SpreadElement).getOperand()
+      or
+      e = ctx.(ForOfStmt).getIterationDomain()
+    )
+  }
+
+  /**
+   * A loop that iterates through some array using the `length` property. 
+   * The loop is either of the style `for(..; i < arr.length;...)` or `while(i < arr.length) {..;i++;..}`. 
+   */
+  class ArrayIterationLoop extends Stmt {
+    LocalVariable indexVariable;
+    LoopStmt loop;
+
+    ArrayIterationLoop() {
+      this = loop and 
+      exists(RelationalComparison compare, DataFlow::PropRead lengthRead |
+        compare = loop.getTest() and
+        compare.getLesserOperand() = indexVariable.getAnAccess() and
+        lengthRead.accesses(_, "length") and
+        lengthRead.flowsToExpr(compare.getGreaterOperand())
+      ) and
+      (
+        loop.(ForStmt).getUpdate().(IncExpr).getOperand() = indexVariable.getAnAccess() or
+        loop.getBody().getAChild*().(IncExpr).getOperand() = indexVariable.getAnAccess()
+      )
+    }
+    
+    /**
+     * Gets the loop test of this loop.
+     */
+    Expr getTest() {
+      result = loop.getTest()
+    }
+    
+    /**
+     * Gets the body of this loop.
+     */
+    Stmt getBody() {
+      result = loop.getBody()   
+    }
+
+    /**
+     * Gets the variable holding the loop variable and current array index.
+     */
+    LocalVariable getIndexVariable() { result = indexVariable }
+  }
+
+  /**
+   * A data flow sink for untrusted user input that is being looped through.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A loop that iterates over an array, such as `for (..; .. sink.length; ...) ...`
+   */
+  private class LoopSink extends Sink {
+    LoopSink() {
+      exists(ArrayIterationLoop loop, Expr lengthAccess, DataFlow::PropRead lengthRead |
+        loop.getTest().(RelationalComparison).getGreaterOperand() = lengthAccess and
+        lengthRead.flowsToExpr(lengthAccess) and
+        lengthRead.accesses(this, "length") and
+        
+        // In the DOS we are looking for arrayRead will evaluate to undefined.
+        // If an obvious nullpointer happens on this undefined, then the DOS cannot happen.
```

Comment:
> If an obvious nullpointer happens on this undefined

It is not really obvious what this means. Maybe just continue the line above with "this may cause an exception to be thrown, thus bailing out of the loop".